language/move-vm/runtime/src/unit_tests/loader_circular.rs

```diff
@@ -0,0 +1,352 @@
+    }
+
+    fn get_resource(
+        &self,
+        _address: &AccountAddress,
+        _tag: &StructTag,
+    ) -> PartialVMResult<Option<Vec<u8>>> {
+        Ok(None)
+    }
+}
+
+fn compile_file(addr: &[u8; AccountAddress::LENGTH]) -> Vec<CompiledModule> {
+    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    path.push("src/unit_tests/modules_circular.move");
+    let s = path.to_str().expect("no path specified").to_owned();
+    let (_, modules) =
+        move_lang::move_compile(&[s], &[], Some(Address::new(*addr))).expect("Error compiling...");
+
+    let mut compiled_modules = vec![];
+    for module in modules {
+        match module {
+            CompiledUnit::Module { module, .. } => compiled_modules.push(module),
+            CompiledUnit::Script { .. } => (),
+        }
+    }
+    compiled_modules
+}
+
+#[test]
+fn all_good() {
+    let data_store = DataStore::empty();
+    let mut adapter = Adapter::new(data_store);
+    let modules = compile_file(&[0; 16]);
+    adapter.publish_modules(modules, "", "");
+    adapter
+        .call_function(
+            &ModuleId::new(WORKING_ACCOUNT, Identifier::new("A").unwrap()),
+            Identifier::new("entry_a").unwrap().as_ident_str(),
+        )
+        .expect("A::entry_a() must work");
+    adapter
+        .call_function(
+            &ModuleId::new(WORKING_ACCOUNT, Identifier::new("B").unwrap()),
+            Identifier::new("entry_b").unwrap().as_ident_str(),
+        )
+        .expect("B::entry_b() must work");
+    adapter
+        .call_function(
+            &ModuleId::new(WORKING_ACCOUNT, Identifier::new("C").unwrap()),
+            Identifier::new("entry_c").unwrap().as_ident_str(),
+        )
+        .expect("C::entry_c() must work");
+    adapter
+        .call_function(
+            &ModuleId::new(WORKING_ACCOUNT, Identifier::new("D").unwrap()),
+            Identifier::new("entry_d").unwrap().as_ident_str(),
+        )
+        .expect("D::entry_d() must work");
+    adapter
+        .call_function(
+            &ModuleId::new(WORKING_ACCOUNT, Identifier::new("E").unwrap()),
+            Identifier::new("entry_e").unwrap().as_ident_str(),
+        )
+        .expect("E::entry_e() must work");
+    adapter
+        .call_function(
+            &ModuleId::new(WORKING_ACCOUNT, Identifier::new("F").unwrap()),
+            Identifier::new("entry_f").unwrap().as_ident_str(),
+        )
+        .expect("F::entry_f() must work");
+}
+
+#[test]
+fn circular_c() {
+    let data_store = DataStore::empty();
+    let mut adapter = Adapter::new(data_store);
+    let modules = compile_file(&[0; 16]);
+    adapter.publish_modules(modules, "C", "A");
+    assert_eq!(
+        adapter
+            .call_function(
+                &ModuleId::new(WORKING_ACCOUNT, Identifier::new("A").unwrap()),
+                Identifier::new("entry_a").unwrap().as_ident_str(),
+            )
+            .unwrap_err(),
+        VMStatus::Error(StatusCode::UNEXPECTED_VERIFIER_ERROR),
```

Comment:
this is reporting a `CIRCULAR_DEPENDENCY` error, but it's obfuscated by the need for `error_specializer` in `execute_function`. I suppose we can rewrite those with `execute_script` which a bit lame but maybe I should...

Reply:
I don't follow.... The execute_function calls in loader_tests.rs are using an identity function for error_specializer. The "obfuscation" seems to be coming from expect_no_verification_errors() in move-vm/runtime/src/loader.rs. What are you suggesting to avoid that?

Reply:
I suppose strictly speaking you are right and my description was not exact.
I guess `execute_function` is returning a `VMStatus` whereas the other functions are returning a `VMError` which has all the info.
So strictly speaking is not the `error_specializer` that is causing that, but having that (the `error_specializer`) is causing us to convert at that level (`into_vm_status` - https://github.com/libra/libra/blob/master/language/move-vm/runtime/src/session.rs#L42) and not being able to see the true error.
Or at least that is my understanding, let me know, if that is not the case (though whether that is the case or not is rather immaterial at this point)

Reply:
Ah, OK, I get it now. Thanks for the explanation